distribution/metadata/v2_metadata_service.go

```diff
@@ -15,10 +18,24 @@ type V2MetadataService struct {
 
 // V2Metadata contains the digest and source repository information for a layer.
 type V2Metadata struct {
+	AuthConfigHash   string
 	Digest           digest.Digest
 	SourceRepository string
 }
 
+// HashAuthConfig returns an sha256 sum of the "authConfig" if successful.
+func HashAuthConfig(authConfig *types.AuthConfig) (string, error) {
+	if authConfig == nil {
+		return "", nil
+	}
+	buf, err := json.Marshal(authConfig)
```

Comment:
I'm wondering if we should marshal a copy of `authConfig` that has the password removed. Two reasons come to mind:

1. Paranoia. If the password is low-complexity, someone might be able to reverse it using the hash of `authConfig`.
2. Currently, password changes would cause a hash mismatch. But we probably don't want them to.